File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import defaultdict

def solution(topping):
    chul = defaultdict(int)
    bro = defaultdict(int)
    
    for t in topping:
        bro[t] += 1
        
    answer =0  
    for i in range(len(topping) - 1):
        chul[topping[i]] += 1
        bro[topping[i]] -= 1
        if bro[topping[i]] == 0:
            del bro[topping[i]]
        
        if len(chul) == len(bro):
            answer += 1
    return answer
        
    
    
    

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        Map<Integer, Integer> rightCounter = new HashMap<>();
        
        for (int t : topping) {
            rightCounter.merge(t, 1, (o,v) -> o + 1);
        }
        
        Map<Integer, Integer> leftCounter = new HashMap<>();
        int answer = 0;
        for (int t : topping) {
            rightCounter.computeIfPresent(t, (k, v) -> v - 1);
            if (rightCounter.get(t) == 0) {
                rightCounter.remove(t);
            }
            leftCounter.merge(t, 1, (o, v) -> o + 1);
            
            if (rightCounter.size() == leftCounter.size()) {
                answer++;
            }
        }
        return answer;
    }
}